rust/ethers-prometheus/src/json_rpc_client.rs

```diff
@@ -0,0 +1,174 @@
+    ///   the request was made on.
+    /// - `method`: request method string.
+    #[builder(setter(into, strip_option), default)]
+    request_duration_seconds: Option<CounterVec>,
+}
+
+/// Expected label names for the metric.
+pub const REQUEST_COUNT_LABELS: &[&str] = &["provider_node", "chain", "method"];
+/// Help string for the metric.
+pub const REQUEST_COUNT_HELP: &str = "Total number of requests made to this client";
+
+/// Expected label names for the metric.
+pub const REQUEST_FAILURE_COUNT_LABELS: &[&str] = &["provider_node", "chain", "method"];
+/// Help string for the metric.
+pub const REQUEST_FAILURE_COUNT_HELP: &str =
+    "Total number of requests made which resulted in an error from the inner client";
+
+/// Expected label names for the metric.
+pub const REQUEST_DURATION_SECONDS_LABELS: &[&str] = &["provider_node", "chain", "method"];
+/// Help string for the metric.
+pub const REQUEST_DURATION_SECONDS_HELP: &str = "Total number of seconds spent making requests";
+
+/// Configuration for the prometheus JsonRpcClioent. This can be loaded via
+/// serde.
+#[derive(Default, Clone, Debug)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[cfg_attr(feature = "serde", serde(tag = "type", rename_all = "camelCase"))]
+pub struct PrometheusJsonRpcClientConfig {
+    /// Information about what node this client is connecting to.
+    pub node: Option<NodeInfo>,
+
+    /// Information about the chain this client is for.
+    pub chain: Option<ChainInfo>,
+}
+
+impl PrometheusJsonRpcClientConfig {
+    fn node_host(&self) -> &str {
+        self.node
+            .as_ref()
+            .and_then(|n| n.host.as_ref())
+            .map(|h| h.as_str())
+            .unwrap_or("unknown")
+    }
+    fn chain_name(&self) -> &str {
+        self.chain
+            .as_ref()
+            .and_then(|c| c.name.as_ref())
+            .map(|n| n.as_str())
+            .unwrap_or("unknown")
+    }
+}
+
+/// An ethers-rs JsonRpcClient wrapper that instruments requests with prometheus
+/// metrics. To make this as flexible as possible, the metric vecs need to be
+/// created and named externally, they should follow the naming convention here
+/// and must include the described labels.
+pub struct PrometheusJsonRpcClient<C> {
+    inner: C,
+    metrics: JsonRpcClientMetrics,
+    config: PrometheusJsonRpcClientConfig,
+}
+
+impl<C> PrometheusJsonRpcClient<C> {
+    /// Wrap a JsonRpcClient with metrics.
+    pub fn new(
+        client: C,
+        metrics: JsonRpcClientMetrics,
+        config: PrometheusJsonRpcClientConfig,
+    ) -> Self {
+        Self {
+            inner: client,
+            metrics,
+            config,
+        }
+    }
+}
+
+impl<C> Debug for PrometheusJsonRpcClient<C>
+where
+    C: JsonRpcClient,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "PrometheusJsonRpcClient({:?})", self.inner)
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
```

Comment:
just curious for my own learning why these are needed

Reply:
Maintains the same conditions on the attribute as the super trait defined in ethers-rs, I don't know a lot about building for wasm but it looks like it has different constraint requirements (well, less requirements actually) and I wanted to make sure we were following the same since it is an implementation of the trait.